### PR TITLE
Improve the facet string distribution speed

### LINKED ISSUE #260: facetsDistribution should return one of the original value
When asking for `facetsDistribution` during the search, the keys are returned lowercased. They should be returned with the case specified by the users. Which is already the behavior of v0.20.0

If the users send two docs with the same value (in a case insensitive way), ex:

```json
[
  { "id": 2, "author": "jane austin" },
  { "id": 456, "author": "Jane Austin" }
]
```

**The behavior should be the same as v0.20.0: undefined.**

## Internal implementation

- if there are more than 10 000 candidates: the value (of the faceted field) of the first document in the DB is returned.

-  if there are less than 10 000  candidates
    - the core engine should NOT return two keys (`jane austin` and `Jane Austin`) but only one.
    - the key chosen is the first document in the DB is returned

-  The return should be the same between 2 calls to the core-engine (with the same search and the same documents)

In any case, the behavior is considered undefined, so the key chosen is not a problem.

### PR DESCRIPTION
**Pull request by Kerollmops**
This pull request creates a data structure similar to the one we use for the faceted numbers, a tetratomic decision tree but this time for the facet strings. This PR also changes the facet distribution behavior by returning one of the original facet values, fixes #260.

This data structure defines bucket-like structures where documents ids are stored under their facet value and helps the search decide if it wants to move to a lower level under a given bucket or not, depending on if the current bucket contains interesting documents or not. The whole format, algorithm, and previous attempts are explained in the [`facet_string.rs` file](https://github.com/meilisearch/milli/blob/ec1cfdd42be570a12fd8d95e27ddee2f0c1332b5/milli/src/search/facet/facet_string.rs).

Note that this data structure **could** be used to sort by string lexicographically, that hypothetically possible. We need more testing, in terms of performance and quality, as we will sort on lowercased versions of the facet values.

 - [x] Implement a faster and more precise way to fetch the facet distribution.
 - [x] Store and return the original facet string value. We currently return the lowercased version.

### CODE REVIEW

File: milli/src/search/facet/facet_distribution.rs [hunk 1]
```diff
@@ -22,7 +23,7 @@ const MAX_VALUES_BY_FACET: usize = 1000;
 
 /// Threshold on the number of candidates that will make
 /// the system to choose between one algorithm or another.
-const CANDIDATES_THRESHOLD: u64 = 1000;
+const CANDIDATES_THRESHOLD: u64 = 3000;
```
[the remark this answers is not in the document]
Comment (Kerollmops) [reply milli/src/search/facet/facet_distribution.rs:26]: Unfortunately, that's a magic value.

File: milli/src/heed_codec/facet/facet_level_value_u32_codec.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+pub struct FacetLevelValueU32Codec;
+
+impl<'a> heed::BytesDecode<'a> for FacetLevelValueU32Codec {
+    type DItem = (FieldId, NonZeroU8, u32, u32);
+
+    fn bytes_decode(bytes: &'a [u8]) -> Option<Self::DItem> {
+        let (field_id_bytes, bytes) = try_split_array_at(bytes)?;
+        let field_id = u16::from_be_bytes(field_id_bytes);
+        let (level, bytes) = bytes.split_first()?;
+        let level = NonZeroU8::new(*level)?;
```
[the remark this answers is not in the document]
Comment (Kerollmops) [reply milli/src/heed_codec/facet/facet_level_value_u32_codec.rs:19]: In terms of correctness, using a `NonZeroU8` helps us be sure that we don't en/decode a zero, which would be invalid.

**Review by ManyTheFish — DISMISSED** [review]

**Comment by ManyTheFish** [comment]
There is a small drawback of this implementation, because we iterate over the facets in Ascendent order. The algorithm will not choose the more popular facets values if there is a lot of available values but choose the first facets string in lexicographic order or the first facet number in ascendant order.

note: this drawback appears when we have a lot of facet values associated with the search candidates.

**Review by ManyTheFish — APPROVED** [review]

**Comment by Kerollmops** [comment]
bors merge

**Pull request merged**